plugins/com.google.cloud.tools.eclipse.appengine.libraries/src/com/google/cloud/tools/eclipse/appengine/libraries/AppEngineLibraryContainerResolverJob.java

```diff
@@ -0,0 +1,242 @@
+                                                                    library, subMonitor.newChild(1));
+          JavaCore.setClasspathContainer(classpathEntry.getPath(), new IJavaProject[] {javaProject},
+                                         new IClasspathContainer[] {container}, null);
+          serializer.saveContainer(javaProject, container);
+        }
+      }
+    } catch (LibraryRepositoryServiceException | CoreException | IOException ex) {
+      return StatusUtil.error(this, "Could not resolve libraries", ex);
+    } finally {
+      releaseRepositoryService();
+    }
+    return Status.OK_STATUS;
+  }
+
+  private LibraryClasspathContainer resolveLibraryFiles(ILibraryRepositoryService repositoryService,
+                                                        IClasspathEntry classpathEntry,
+                                                        Library library,
+                                                        IProgressMonitor monitor) 
+                                                            throws CoreException, LibraryRepositoryServiceException {
+    List<LibraryFile> libraryFiles = library.getLibraryFiles();
+    SubMonitor subMonitor = SubMonitor.convert(monitor, libraryFiles.size());
+    subMonitor.subTask("Resolving artifacts for " + getLibraryDescription(library));
+    SubMonitor child = subMonitor.newChild(libraryFiles.size());
+
+    IClasspathEntry[] entries = new IClasspathEntry[libraryFiles.size()];
+    int idx = 0;
+    for (LibraryFile libraryFile : libraryFiles) {
+      IClasspathAttribute[] libraryFileClasspathAttributes = getClasspathAttributes(libraryFile);
+      entries[idx++] =
+          JavaCore.newLibraryEntry(repositoryService.getJarLocation(libraryFile.getMavenCoordinates()),
+                                   getSourceLocation(repositoryService, libraryFile),
+                                   null,
+                                   getAccessRules(libraryFile.getFilters()),
+                                   libraryFileClasspathAttributes,
+                                   true);
+      child.worked(1);
+    }
+    monitor.done();
+    LibraryClasspathContainer container = new LibraryClasspathContainer(classpathEntry.getPath(),
+                                                                        getLibraryDescription(library),
+                                                                        entries);
+    return container;
+  }
+
+  private static int getTotalwork(IClasspathEntry[] rawClasspath) {
+    int sum = 0;
+    for (IClasspathEntry element : rawClasspath) {
+      if (isLibraryClasspathEntry(element.getPath())) {
+        ++sum;
+      }
+    }
+    return sum;
+  }
+
+  private static boolean isLibraryClasspathEntry(IPath path) {
+    return path != null && path.segmentCount() == 2 && Library.CONTAINER_PATH_PREFIX.equals(path.segment(0));
+  }
+
+  private static String getLibraryDescription(Library library) {
```

Comment:
I think this is the second place I've seen you do this for `Library` which suggests that either this should be in `Library`, or perhaps a `ILabelProvider`.  And perhaps `getLabel()`?  _description_ suggests a longer detailed description.